tools/wad/wad.1
.\"                              hey, Emacs:   -*- nroff -*-
.\" wad is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.\"
.\" See the GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; see the file COPYING.  If not, write to:
.\"
.\"		Free Software Foundation, Inc.
.\"		59 Temple Place, Suite 330
.\"		Boston, MA 02111-1307, USA
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins (default)
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.\"
.TH WAD 1 "9 Jan, 2004" QuakeForge "QuakeForge User's Manual"
.\" Please update the above date whenever this man page is modified.
.SH NAME
wad \- The QuakeForge Packfile Tool
.SH SYNOPSIS
.B wad
<\fBcommand\fP> [\fBoptions\fP] \fIARCHIVE\fP \fIFILE\fP...
.SH DESCRIPTION
\fBwad\fP is a utility for managing pack (*.wad) archives. The format used is
compatible with Quake and Quake II. wad archives are a convenient way to store
groups of related files. Some engines allow wad archives to be compressed
internally.
.SH COMMANDS
\fBwad\fP accepts the following arguments as a command. Only one command may be
used at a time.
.TP
.B \-c, \-\-create
Create an archive. Overwrites any contents the archive may have had.
.TP
.B \-t, \-\-test
Test an archive for valid formatting.
.TP
.B \-x, \-\-extract
Extract an archive. Overwrites any files in the current directory with the same
names as those inside the archive.
.TP
.B \-h, \-\-help
Show summary of options.
.TP
.B \-V, \-\-version
Show the version of wad.
.SH OPTIONS
\fBwad\fP takes the following arguments as options.
.TP
.B \-f, \-\-file ARCHIVE
Use ARCHIVE as the archive filename, instead of the first non-option argument.
This option is provided for compatibility with older versions of \fBwad\fP.
.TP
.B \-p, \-\-pad
Can be useful when creating an archive. File space is padded to a 4\-byte
boundary, assisting in the speed of loading files from the archive. The file
data itself is not changed.
.TP
.B \-q, \-\-quiet
Inhibit some of wad's normal output.
.TP
.B \-v, \-\-verbose
Display more output than usual.
.SH "EXIT STATUS"
\fBwad\fP returns a zero exit status when it has completed a command
successfully (without error). Otherwise, it returns a non-zero exit status.
.SH BUGS
.PP
The program does not currently handle compression or decompression. This
is planned for a future version.
.PP
The program does not currently handle extraction of individual files inside an
archive. This is planned for a future version.
.PP
The program cannot modify an archive once it has been created.
.SH AUTHORS
Bill Currie (taniwha@quakeforge.net) wrote most of the program.
.PP
Jeff Teunissen (deek@quakeforge.net) wrote the command-line interface and the
documentation.
.SH "SEE ALSO"
.BR quakeforge (1),
.BR qfcc (1)


NEWS
NEWS for the QuakeForge project
-------------------------------

Changes from 0.5.2

    o General enhancements.
        * MVD playback in the qw clients.
        * MVD recording in the qw server.

Changes from 0.3.0

    o General enhancements.
        * NQ (quake1, single player, NetQuake, WinQuake) is back.
        * Video (driver and renderer), audio, gamecode (progs), console,
          model loading and general utility code shared in common libs between
          NQ and QW
        * Plugin system for sound and cd drivers.
        * Ogg Vorbis compression for sound effect samples (streamed ogg support
          is planned).
        * Menu system re-written in QuakeC/Ruamoko
        * Curses console for servers in Unix.
        * New keybinding system.

    o Progs engine enhancements.
        * Fully modular. Mod/Server-independant, allowing for
          stand-alone QuakeC interpreter and client-side code.
        * The engine is no longer dependent on progdefs.h. This makes
          QuakeForge have trouble with QuakeC compilers that strip
          symbol names -- but those can be repaired using the new
          "qfdefs" tool.
        * New basic types, and instructions for using them, for version
          0.fff.002 progs (Quake's Version 6 is still supported).
        * Better dynamic string handling. Dynamic strings are
          garbage-collected.
        * Object-oriented runtime system, in the style of Objective-C.
        * Runtime fixups of built-in functions whose builtin number is
          zero. If a built-in function is available to the engine, with
          the same name given in the source code (e.g.
          "void (void) coredump = #0;"), the engine will set the value
          to the actual number used by the engine.
        * Debugging support (needs compiler support, which qfcc
          provides), including line-number information. Can display the
          line of text in the source on which an error occurred, if the
          new Cvar "pr_source_path" is set correctly.
        * Instruction-level code dumps are now formatted in an
          easier-to-read "assembly-code" format.

    o QuakeForge Code Compiler, qfcc -- a replacement for qcc.
        * Uses libQFutil for general utility functions such as hash
          tables, file access (incomplete), and pak files.
        * Uses libQFgamecode for progs information, such as opcode
          tables, enums, types, etc.)
        * Uses a lex/yacc (actually, Flex and Bison) scanner/parser
          architecture, like a "real" compiler does.
        * Drastically reduces the number of global defs needed for
          a given progs, by reusing temporary globals when possible.
        * All functions use the same block of defs for their local
          variables (compatible with all known servers), which further
          reduces the number of global defs needed.
        * Nested/chained function calls work properly.
        * Simple expressions, such as "a = b + c", compile to a single
          instruction instead of two, and do not use a temporary
          variable.
        * Local variables can be initialized to known values, as in
          "local float foo = 0.1;".
        * Uses the C preprocessor, so header files are now possible and
          useful.
        * The compiler checks for uninitialized and unused variables in
          functions, to help in finding errors.
        * The compiler can generate and use object files, for separate
          compilation (you don't have to compile all of your source all
          the time, only the parts that have changed). Normally, object
          files have a .qfo extension.
        * The compiler can link libraries, which are implemented as .qfo
          files inside pak archives.
        * The compiler implements stricter type checking, to assist in
          finding and fixing code errors.
        * The C language's "for", "do...while", and "switch" control
          structures have been adopted to provide a richer set of
          structures for programmers to use.
        * Additional functionality is given for progs that are to be
          used with QuakeForge's more advanced progs engine:

            x IMPORTANT:
              The order of operations is slightly different in "native"
              mode. The evaluation of "!foo & bar" is no longer
              "!(foo & bar)", but rather "(!foo) & bar". The new order of
              operations rules is more correct, but is different from how
              QCC evaluated. If you give the --traditional switch to
              qfcc's command-line, the old behavior is used.
            x The compiler now allows you to create your own functions
              that can accept a variable number of arguments, using the
              special "..." notation. The special constants "@argc" and
              "@argv" help with this. @argc contains the number of
              arguments, and @argv contains a list of those arguments. The
              handling of this is an improvement on C's vararg function
              handling, in our opinion, made possible by how the Quake
              system works.
            x New basic types: integer, pointer, enum, and id.
            x You can create new complex types, like arrays, structures,
              and unions, and allocate them at runtime.
            x String operators. You can add strings together to
              concatenate them, and you can compare their contents using
              the <=, <, >, >= operators.
            x The integer type has the full array of C operations
              available to it.
            x QFCC and QuakeForge now contain an object-oriented
              programming system, in the style of Objective-C and
              Smalltalk. It's *very* powerful, and can be used to create
              lots of interesting new things with rather little effort. If
              you don't wish to use the new system, you can safely ignore

libs/util/wad.c
/*
	wad.c

	(description)

	Copyright (C) 1996-1997  Id Software, Inc.

	This program is free software; you can redistribute it and/or
	modify it under the terms of the GNU General Public License
	as published by the Free Software Foundation; either version 2
	of the License, or (at your option) any later version.

	This program is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

	See the GNU General Public License for more details.

	You should have received a copy of the GNU General Public License
	along with this program; if not, write to:

		Free Software Foundation, Inc.
		59 Temple Place - Suite 330
		Boston, MA  02111-1307, USA

*/
#ifdef HAVE_CONFIG_H
# include "config.h"
#endif

#ifdef HAVE_STRING_H
# include <string.h>
#endif
#ifdef HAVE_STRINGS_H
# include <strings.h>
#endif

#include "QF/cvar.h"
#include "QF/qendian.h"
#include "QF/quakefs.h"
#include "QF/sys.h"
#include "QF/wad.h"
#include "QF/zone.h"

int         wad_numlumps;
lumpinfo_t *wad_lumps;
byte       *wad_base;

/*
	W_CleanupName

	Lowercases name and pads with spaces and a terminating 0 to the length of
	lumpinfo_t->name. Used so lumpname lookups can proceed rapidly by
	comparing 4 chars at a time Space padding is so names can be printed
	nicely in tables. Can safely be performed in place.
*/
static void
W_CleanupName (const char *in, char *out)
{
	int			i;
	int			c;

	for (i = 0; i < 16; i++) {
		c = in[i];
		if (!c)
			break;

		if (c >= 'A' && c <= 'Z')
			c += ('a' - 'A');
		out[i] = c;
	}

	for (; i < 16; i++)
		out[i] = 0;
}

typedef struct wad_alloc_s {
	memhunk_t *hunk;
	const char *name;
} wad_alloc_t;

static void *
wad_allocator (void *data, size_t size)
{
	wad_alloc_t *alloc = data;
	return Hunk_RawAllocName (alloc->hunk, size, alloc->name);
}

VISIBLE void
W_LoadWadFile (const char *filename, memhunk_t *hunk)
{
	lumpinfo_t *lump_p;
	wadinfo_t  *header;
	int         i;
	int			infotableofs;

	wad_alloc_t wad_alloc = {
		.hunk = hunk,
		.name = filename,
	};
	qfs_allocator_t alloc = {
		.alloc = wad_allocator,
		.data = &wad_alloc,
	};
	wad_base = QFS_LoadFile (QFS_FOpenFile (filename), &alloc);
	if (!wad_base)
		Sys_Error ("W_LoadWadFile: unable to load %s", filename);

	header = (wadinfo_t *) wad_base;

	if (header->id[0] != 'W' || header->id[1] != 'A' || header->id[2] != 'D'
		|| header->id[3] != '2')
		Sys_Error ("Wad file %s doesn't have WAD2 id", filename);

	wad_numlumps = LittleLong (header->numlumps);
	infotableofs = LittleLong (header->infotableofs);
	wad_lumps = (lumpinfo_t *) (wad_base + infotableofs);

	for (i = 0, lump_p = wad_lumps; i < wad_numlumps; i++, lump_p++) {
		lump_p->filepos = LittleLong (lump_p->filepos);
		lump_p->size = LittleLong (lump_p->size);
		W_CleanupName (lump_p->name, lump_p->name);
		if (lump_p->type == TYP_QPIC)
			SwapPic ((qpic_t *) (wad_base + lump_p->filepos));
	}
}

static lumpinfo_t *
W_GetLumpinfo (const char *name)
{
	int         i;
	lumpinfo_t *lump_p;
	char        clean[16];

	W_CleanupName (name, clean);

	for (lump_p = wad_lumps, i = 0; i < wad_numlumps; i++, lump_p++) {
		if (!strcmp (clean, lump_p->name))
			return lump_p;
	}

	Sys_MaskPrintf (SYS_warn, "W_GetLumpinfo: %s not found", name);
	return NULL;
}

VISIBLE void *
W_GetLumpName (const char *name)
{
	lumpinfo_t *lump;

	if (!wad_base) {
		return 0;
	}
	lump = W_GetLumpinfo (name);

	return lump ? (wad_base + lump->filepos) : 0;
}

VISIBLE void
SwapPic (qpic_t *pic)
{
	pic->width = LittleLong (pic->width);
	pic->height = LittleLong (pic->height);
}
